src/metabase/query_processor/middleware/auto_bucket_datetimes.clj

```diff
@@ -97,56 +98,68 @@
    x]
   (cond
-    ;; do not autobucket Fields in a non-compound filter clause that either:
+    ;; do not autobucket clauses in a non-compound filter clause that either:
     (simple-filter-clause? query stage-path x)
     (cond
       ;; *  is not an equality or comparison filter. e.g. wouldn't make sense to bucket a field and then check if it is
       ;;    `NOT NULL`
       (not (mbql.u/is-clause? #{:= :!= :< :> :<= :>= :between} x))
       :do-not-bucket-reason/not-equality-or-comparison-filter
 
-      ;; *  has arguments that aren't `yyyy-MM-dd` date strings. The only reason we auto-bucket datetime Fields in the
+      ;; *  has arguments that aren't `yyyy-MM-dd` date strings. The only reason we auto-bucket datetime clauses in the
       ;; *  first place is for legacy reasons, if someone is specifying additional info like hour/minute then we
       ;; *  shouldn't assume they want to bucket by day
-      (let [[_tag _opts _field & values] x]
+      (let [[_tag _opts _ref & values] x]
         (not (every? auto-bucketable-value? values)))
       :do-not-bucket-reason/not-all-values-are-auto-bucketable)
 
-    ;; *  do not autobucket fields that are updating the time interval
+    ;; *  do not autobucket clauses that are updating the time interval
     (lib.util.match/match-one x
-      [(_tag :guard #{:+ :-}) _ [:field _ _] [:interval _ _n (unit :guard #{:minute :hour :second})]])
+      [(_tag :guard #{:+ :-})
+       _
+       [(_ :guard #{:expression :field}) _ _]
+       [:interval _ _n (unit :guard #{:minute :hour :second})]])
     :do-not-bucket-reason/bucket-between-relative-starting-from
 
-    ;; do not auto-bucket fields inside a `:time-interval` filter: it already supplies its own unit
-    ;; do not auto-bucket fields inside a `:datetime-diff` clause: the precise timestamp is needed for the difference
+    ;; do not auto-bucket clauses inside a `:time-interval` filter: it already supplies its own unit
+    ;; do not auto-bucket clauses inside a `:datetime-diff` clause: the precise timestamp is needed for the difference
     (mbql.u/is-clause? #{:time-interval :datetime-diff} x)
     :do-not-bucket-reason/bucketed-or-precise-operation
 
-    ;; do not autobucket Fields that already have a temporal unit, or have a binning strategy
-    (and (mbql.u/is-clause? :field x)
+    ;; do not autobucket clauses that already have a temporal unit, or have a binning strategy
+    (and (or (mbql.u/is-clause? :expression x)
+             (mbql.u/is-clause? :field x))
          (let [[_tag opts _id-or-name] x]
            ((some-fn :temporal-unit :binning) opts)))
     :do-not-bucket-reason/field-with-bucketing-or-binning))
 
-(mu/defn ^:private date-or-datetime-field?
+(mu/defn ^:private date-or-datetime-clause?
   [{base-type :base-type, effective-type :effective-type} :- ::field-type-info]
+  (when (and (nil? base-type) (nil? effective-type))
+    (log/warn "`type-info` is missing `base-type` and `effective-type`."))
```

Comment:
How common is this situation? Do we really want our users to read this message?

Reply:
I believe this should not happen. Would changing the log level to `DEBUG` be the right approach?

Reply:
I agree it should not happen, yet we have so much legacy (rubbish) in databases, that my bet would be that it does happen. And then interesting questions are
- is this code on a path that runs often enough that it can produce substantial spam? are around logs that we know get written just as frequently, i.e., can we be sure we won't add much more spam?
- is this information crucial for debugging errors that are difficult to reproduce?

Reply:
Thinking about it again, I've decided to remove the log completely. It could be helpful, but if we know that some column is missing temporal-unit, there is not that many places one should be looking into.